core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java

```diff
@@ -165,20 +175,55 @@ private void setTelemetryModules(ApplicationInsightsXmlConfiguration appConfigur
     }
 
     /**
-     * Setting an instrumentation key
-     * @param instrumentationKey The instrumentation key found in the configuration.
+     * Setting an instrumentation key:
+     * First we try the JVM option '-DAPPLICATION_INSIGHTS_IKEY=i_key'
+     * Next we will try the environment variable 'APPLICATION_INSIGHTS_IKEY',
+     * Next we will try to fetch the i-key from the ApplicationInsights.xml
+     * @param userConfiguration The configuration that was represents the user's configuration in ApplicationInsights.xml.
      * @param configuration The configuration class.
      * @return True if succeeded.
      */
-    private boolean setInstrumentationKey(String instrumentationKey, TelemetryConfiguration configuration) {
+    private boolean setInstrumentationKey(ApplicationInsightsXmlConfiguration userConfiguration, TelemetryConfiguration configuration) {
         try {
-            configuration.setInstrumentationKey(instrumentationKey);
+            // First, try to find the i-key as an Env variable 'APPLICATION_INSIGHTS_IKEY'
+            RuntimeMXBean runtimeMxBean = ManagementFactory.getRuntimeMXBean();
+            List<String> arguments = runtimeMxBean.getInputArguments();
+            if (arguments != null) {
+                for (String argument : arguments) {
+                    String[] ikeyData = argument.split("=");
+                    if (ikeyData.length != 2) {
+                        continue;
+                    }
 
-            return true;
+                    if (!JVM_VARIABLE_I_KEY.equals(ikeyData[0])) {
+                        continue;
+                    }
+
+                    String ikey = ikeyData[1];
+                    if (!Strings.isNullOrEmpty(ikey)) {
+                        configuration.setInstrumentationKey(argument);
+                        return true;
+                    }
+                }
+            }
+
+            // Second, try to find the i-key as an JVM argument '-DAPPLICATION_INSIGHTS_IKEY=i_key'
+            String ikey = System.getenv(ENV_VARIABLE_I_KEY);
+            if (!Strings.isNullOrEmpty(ikey)) {
+                configuration.setInstrumentationKey(ikey);
+                return true;
+            }
+
+            // Else, try to find the i-key in the ApplicationInsights.xml
+            if (userConfiguration != null) {
+                configuration.setInstrumentationKey(userConfiguration.getInstrumentationKey());
```

Comment:
I will change that line and will also trim the value and will only assign if the configuration is not null and not empty